src/Interactive/Features/Interactive/Core/InteractiveHost.RemoteService.cs

```diff
@@ -13,15 +14,19 @@ internal partial class InteractiveHost
     {
         internal sealed class RemoteService
         {
+            private const int ProcessExitHooked = 1;
+            private const int ProcessExitHandled = 2;
+
             public readonly Process Process;
             public readonly Service Service;
             private readonly int _processId;
+            private SemaphoreSlim _disposeSemaphore = new SemaphoreSlim(initialCount: 1);
 
             // output pumping threads (stream output from stdout/stderr of the host process to the output/errorOutput writers)
             private Thread _readOutputThread;           // nulled on dispose
             private Thread _readErrorOutputThread;      // nulled on dispose
             private InteractiveHost _host;              // nulled on dispose
-            private bool _disposing;                    // set to true on dispose
+            private volatile int _processExitHandling;  // set to ProcessExitHandled on dispose
```

Comment:
>int [](start = 29, length = 3)

Sorry, one more thing: now that we don't use Interlocked we could define the ProcessExitXXX values as an enum (ProcessExitHandlerStatus { Uninitialized = 0, Hooked = 1, Handled = 2 } ). Would make the code more readable.

Reply:
Good idea! Thank you!

---
In reply to: [163422405](https://github.com/dotnet/roslyn/pull/24058#discussion_r163422405) [](ancestors = 163422405)